File: legacy/avalon-sandbox/avalon-net/Composition/ModelDefault/DefaultSystemContext.cs
/* 
* Copyright 2003-2004 The Apache Software Foundation
* Licensed  under the  Apache License,  Version 2.0  (the "License");
* you may not use  this file  except in  compliance with the License.
* You may obtain a copy of the License at 
* 
*   http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed  under the  License is distributed on an "AS IS" BASIS,
* WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
* implied.
* 
* See the License for the specific language governing permissions and
* limitations under the License.
*/
namespace Apache.Avalon.Composition.Model.Default
{
	using System;

	using Apache.Avalon.Framework;
	using Apache.Avalon.Meta;
	using Apache.Avalon.Repository;
	using Apache.Avalon.Composition.Data;
	using Apache.Avalon.Composition.Logging;
	using Apache.Avalon.Composition.Model;
	
	/// <summary> Implementation of a system context that exposes a system wide set of parameters.
	/// 
	/// </summary>
	/// <author>  <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
	/// </author>
	/// <version>  $Revision: 1.2 $ $Date: 2004/02/29 18:07:17 $
	/// </version>
	public class DefaultSystemContext : Apache.Avalon.Framework.DefaultContext, ISystemContext
	{
		//--------------------------------------------------------------
		// static
		//--------------------------------------------------------------
		
		//--------------------------------------------------------------
		// immutable state
		//--------------------------------------------------------------
		
		private System.IO.FileInfo m_base;
		
		private System.IO.FileInfo m_home;
		
		private System.IO.FileInfo m_temp;
		
		private IRepository m_repository;
		
		private ILoggingManager m_logging;
		
		private ILogger m_logger;
		
		private IModelFactory m_factory;
		
		private IRuntime m_runtime;
		
		private long m_timeout;
		
		private bool m_secure;
		
		//--------------------------------------------------------------
		// mutable state
		//--------------------------------------------------------------
		
		private bool m_trace;
		
		//--------------------------------------------------------------
		// constructor
		//--------------------------------------------------------------
		
		/// <summary> Creation of a new system context.
		/// 
		/// </summary>
		/// <param name="context">a repository initial context
		/// </param>
		/// <param name="artifact">an artifact identifying the default runtime
		/// </param>
		/// <param name="logging">the logging manager
		/// </param>
		/// <param name="base">the base directory from which relative references 
		/// within a classpath or library directive shall be resolved
		/// </param>
		/// <param name="home">the home directory
		/// </param>
		/// <param name="temp">the temp directory
		/// </param>
		/// <param name="repository">the application repository to be used when resolving 
		/// resource directives
		/// </param>
		/// <param name="category">the kernel logging category name
		/// </param>
		/// <param name="trace">flag indicating if internal logging is enabled
		/// </param>
		/// <param name="timeout">a system wide default deployment timeout
		/// </param>
		/*public DefaultSystemContext(Artifact artifact, ILoggingManager logging, 
			System.IO.FileInfo base_Renamed, System.IO.FileInfo home, 
			System.IO.FileInfo temp, IRepository repository, 
			String category, bool trace, long timeout, 
			bool secure) : 
			this(artifact, null, logging, base_Renamed, home, temp, 
			repository, category, trace, timeout, secure)
		{
		}*/
		
		/// <summary> Creation of a new system context.
		/// 
		/// </summary>
		/// <param name="runtime">the runtime instance
		/// </param>
		/// <param name="logging">the logging manager
		/// </param>
		/// <param name="base">the base directory from which relative references 
		/// within a classpath or library directive shall be resolved
		/// </param>
		/// <param name="home">the home directory
		/// </param>
		/// <param name="temp">the temp directory
		/// </param>
		/// <param name="repository">the application repository to be used when resolving 
		/// resource directives
		/// </param>
		/// <param name="category">the kernel logging category name
		/// </param>
		/// <param name="trace">flag indicating if internal logging is enabled
		/// </param>
		/// <param name="timeout">a system wide default deployment timeout
		/// </param>
		/*public DefaultSystemContext(IRuntime runtime, ILoggingManager logging, 
			System.IO.FileInfo base_Renamed, System.IO.FileInfo home, System.IO.FileInfo temp, 
			IRepository repository, String category, bool trace, long timeout, bool secure) : 
			this(runtime, logging, base_Renamed, home, temp, repository, category, trace, timeout, secure)
		{
		}*/
		
		/// <summary> Creation of a new system context.
		/// 
		/// </summary>
		/// <param name="context">the repository intial context
		/// </param>
		/// <param name="runtime">the runtime instance
		/// </param>
		/// <param name="runtime">the runtime class
		/// </param>
		/// <param name="logging">the logging manager
		/// </param>
		/// <param name="base">the base directory from which relative references 
		/// within a classpath or library directive shall be resolved
		/// </param>
		/// <param name="home">the home directory
		/// </param>
		/// <param name="temp">the temp directory
		/// </param>
		/// <param name="repository">the application repository to be used when resolving 
		/// resource directives
		/// </param>
		/// <param name="category">the kernel logging category name
		/// </param>
		/// <param name="trace">flag indicating if internal logging is enabled
		/// </param>
		/// <param name="timeout">a system wide default deployment timeout
		/// </param>
		public DefaultSystemContext(IRuntime runtime,  
			ILoggingManager logging, System.IO.FileInfo base_Renamed, System.IO.FileInfo home, 
			System.IO.FileInfo temp, IRepository repository, String category, bool trace, 
			long timeout, bool secure)
		{
			if (base_Renamed == null)
			{
				throw new System.ArgumentNullException("base");
			}
			/*
			if (repository == null)
			{
				throw new System.ArgumentNullException("repository");
			}*/
			if (logging == null)
			{
				throw new System.ArgumentNullException("logger");
			}
			if (!System.IO.Directory.Exists(base_Renamed.FullName))
			{
				String error = "system.error.base-not-a-directory"  + " " +  base_Renamed.FullName;
				throw new System.ArgumentException(error);
			}
			
			m_base = base_Renamed;
			m_home = home;
			m_temp = temp;
			m_trace = trace;
			m_repository = repository;
			m_logging = logging;
			m_timeout = timeout;
			m_secure = secure;
			m_runtime = runtime;
			
			m_logger = m_logging.GetLoggerForCategory(category);
			m_factory = new StandardModelFactory(this);
		}
		
		//--------------------------------------------------------------
		// ISystemContext
		//--------------------------------------------------------------
		
		/// <summary> Return the model factory.
		/// 
		/// </summary>
		/// <returns> the factory
		/// </returns>
		public virtual IModelFactory ModelFactory
		{
			get
			{
				return m_factory;
			}
		}
		
		/// <summary> Return the base directory from which relative classloader 
		/// references may be resolved.
		/// 
		/// </summary>
		/// <returns> the base directory
		/// </returns>
		public virtual System.IO.FileInfo BaseDirectory
		{
			get
			{
				return m_base;
			}
		}
		
		/// <summary> Return the working directory from which containers may 
		/// establish persistent content.
		/// 
		/// </summary>
		/// <returns> the working directory
		/// </returns>
		public virtual System.IO.FileInfo HomeDirectory
		{
			get
			{
				return m_home;
			}
		}
		
		/// <summary> Return the temporary directory from which a container 
		/// may use to establish a transient content directory. 
		/// 
		/// </summary>
		/// <returns> the temporary directory
		/// </returns>
		public virtual System.IO.FileInfo TempDirectory
		{
			get
			{
				return m_temp;
			}
		}
		
		/// <summary> Return the system wide repository from which resource 
		/// directives can be resolved.
		/// 
		/// </summary>
		/// <returns> the repository
		/// </returns>
		public virtual IRepository Repository
		{
			get
			{
				return m_repository;
			}
		}
		
		/// <summary> Return the system trace flag.
		/// 
		/// </summary>
		/// <returns> the trace flag
		/// </returns>
		public virtual bool IsTraceEnabled
		{
			get
			{
				return m_trace;
			}
		}
		
		/// <summary> Set the system trace flag.
		/// 
		/// </summary>
		/// <param name="trace">the trace flag
		/// </param>
		public virtual void SetTraceEnabled(bool trace)
		{
			m_trace = trace;
		}
		
		/// <summary> Return the logging manager.
		/// 
		/// </summary>
		/// <returns> the logging manager.
		/// </returns>
		public virtual ILoggingManager LoggingManager
		{
			get
			{
				return m_logging;
			}
		}
		
		/// <summary> Return the system logging channel.
		/// 
		/// </summary>
		/// <returns> the system logging channel
		/// </returns>
		public virtual ILogger Logger
		{
			get
			{
				return m_logger;
			}
		}
		
		/// <summary> Return the default deployment phase timeout value.</summary>
		/// <returns> the timeout value
		/// </returns>
		public virtual long DefaultDeploymentTimeout
		{
			get
			{
				return m_timeout;
			}
		}
		
		/// <summary> Return the enabled status of the code security policy.</summary>
		/// <returns> the code security enabled status
		/// </returns>
		/*public virtual bool isCodeSecurityEnabled()
		{
			return m_secure;
		}*/
		
		//------------------------------------------------------------------
		// runtime operations
		//------------------------------------------------------------------
		
		/// <summary> Request the commissioning of a runtime for a supplied deployment 
		/// model.
		/// </summary>
		/// <param name="model">the deployment model 
		/// </param>
		/// <exception cref=""> Exception of a commissioning error occurs
		/// </exception>
		public virtual void Commission(IDeploymentModel model)
		{
			Runtime.Commission(model);
		}
		
		/// <summary> Request the decommissioning of a runtime for a supplied deployment 
		/// model.
		/// </summary>
		/// <param name="model">the deployment model 
		/// </param>
		/// <exception cref=""> Exception of a commissioning error occurs
		/// </exception>
		public virtual void Decommission(IDeploymentModel model)
		{
			Runtime.Decommission(model);
		}
		
		/// <summary> Request resolution of an object from the runtime.</summary>
		/// <param name="model">the deployment model
		/// </param>
		/// <exception cref=""> Exception if a deployment error occurs
		/// </exception>
		public virtual System.Object Resolve(IDeploymentModel model)
		{
			return Runtime.Resolve(model);
		}
		
		/// <summary> Request the release of an object from the runtime.</summary>
		/// <param name="model">the deployment model
		/// </param>
		/// <param name="instance">the object to release
		/// </param>
		/// <exception cref=""> Exception if a deployment error occurs
		/// </exception>
		public virtual void Release(IDeploymentModel model, System.Object instance)
		{
			Runtime.Release(model, instance);
		}

		/// <summary> Return the runtime factory.
		/// 
		/// </summary>
		/// <returns> the factory
		/// </returns>
		private IRuntime Runtime
		{
			get
			{
				return m_runtime;
			}
			
		}
		
		//------------------------------------------------------------------
		// runtime operations
		//------------------------------------------------------------------
		
		/// <summary> Get the runtime class referenced by the artifact.</summary>
		/// <param name="context">the repository initial context
		/// </param>
		/// <param name="artifact">the factory artifact
		/// </param>
		/// <returns> the Runtime class
		/// </returns>
		/*private System.Type GetRuntimeClass(InitialContext context, Artifact artifact)
		{
			if (null == artifact)
				return null;
			
			try
			{
				ClassLoader classloader = typeof(DefaultSystemContext).getClassLoader();
				Builder builder = context.newBuilder(classloader, artifact);
				return builder.getFactoryClass();
			}
			catch (System.Exception e)
			{
				String error = "todo: error message"; //"system.error.load", artifact.toString());
				throw new SystemException(error, e);
			}
		}*/
		
		/// <summary> Build a runtime using a supplied class.
		/// 
		/// </summary>
		/// <param name="type">the log target factory class
		/// </param>
		/// <returns> a instance of the class
		/// </returns>
		/// <exception cref=""> SystemException if the class does not expose a public 
		/// constructor, or the constructor requires arguments that the 
		/// builder cannot resolve, or if a unexpected instantiation error 
		/// ooccurs
		/// </exception>
		/*
		public virtual Runtime buildRuntimeInstance(InitialContext context, System.Type type)
		{
			if (null == type)
				return null;
			
			System.Reflection.ConstructorInfo[] constructors = type.GetConstructors();
			if (constructors.Length < 1)
			{
				String error = "todo: error message"; //"system.error.runtime.no-constructor", type.FullName);
				throw new SystemException(error);
			}
			
			//
			// log target factories only have one constructor
			//
			
			System.Reflection.ConstructorInfo constructor = constructors[0];
			System.Type[] classes = constructor.GetParameters();
			System.Object[] args = new System.Object[classes.Length];
			for (int i = 0; i < classes.Length; i++)
			{
				System.Type c = classes[i];
				if (typeof(ISystemContext).IsAssignableFrom(c))
				{
					args[i] = this;
				}
				else if (typeof(InitialContext).IsAssignableFrom(c))
				{
					args[i] = context;
				}
				else
				{
					String error = "todo: error message"; //"system.error.unrecognized-runtime-parameter", c.FullName, type.FullName);
					throw new SystemException(error);
				}
			}
			
			//
			// instantiate the factory
			//
			
			return instantiateRuntime(constructor, args);
		}*/
		
		/// <summary> Instantiation of a runtime using a supplied constructor 
		/// and arguments.
		/// 
		/// </summary>
		/// <param name="constructor">the runtime constructor
		/// </param>
		/// <param name="args">the constructor arguments
		/// </param>
		/// <returns> the runtime instance
		/// </returns>
		/// <exception cref=""> SystemException if an instantiation error occurs
		/// </exception>
		/*
		private Runtime instantiateRuntime(System.Reflection.ConstructorInfo constructor, System.Object[] args)
		{
			System.Type type = constructor.DeclaringType;
			try
			{
				return (Runtime) constructor.newInstance(args);
			}
			catch (System.Exception e)
			{
				String error = "todo: error message"; //"system.error.runtime-instantiation", type.FullName);
				throw new SystemException(error, e);
			}
		}*/
	}
}

File: avalon/composition/impl/src/java/org/apache/avalon/composition/model/impl/DefaultSystemContext.java
/* 
 * Copyright 2004 Apache Software Foundation
 * Licensed  under the  Apache License,  Version 2.0  (the "License");
 * you may not use  this file  except in  compliance with the License.
 * You may obtain a copy of the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed  under the  License is distributed on an "AS IS" BASIS,
 * WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
 * implied.
 * 
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avalon.composition.model.impl;

import java.io.File;
import java.util.Map;
import java.lang.reflect.Constructor;

import org.apache.avalon.logging.provider.LoggingManager;

import org.apache.avalon.composition.data.SecurityProfile;
import org.apache.avalon.composition.model.DeploymentModel;
import org.apache.avalon.composition.model.impl.DefaultModelFactory;
import org.apache.avalon.composition.provider.ModelFactory;
import org.apache.avalon.composition.provider.SystemContext;
import org.apache.avalon.composition.provider.SystemException;
import org.apache.avalon.composition.provider.Runtime;

import org.apache.avalon.repository.Artifact;
import org.apache.avalon.repository.Repository;
import org.apache.avalon.repository.provider.InitialContext;
import org.apache.avalon.repository.provider.Builder;

import org.apache.avalon.framework.logger.Logger;
import org.apache.avalon.framework.context.Context;
import org.apache.avalon.framework.context.DefaultContext;

import org.apache.avalon.util.i18n.ResourceManager;
import org.apache.avalon.util.i18n.Resources;

/**
 * Implementation of a system context that exposes a system wide set of parameters.
 *
 * @author <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
 * @version $Revision: 1.32 $ $Date: 2004/05/09 23:51:08 $
 */
public class DefaultSystemContext extends DefaultContext 
  implements SystemContext
{
    //--------------------------------------------------------------
    // static
    //--------------------------------------------------------------

    private static final Resources REZ =
            ResourceManager.getPackageResources( DefaultSystemContext.class );

    //--------------------------------------------------------------
    // immutable state
    //--------------------------------------------------------------

    private final File m_base;

    private final File m_home;

    private final File m_temp;

    private final File m_anchor;

    private final Repository m_repository;

    private final ClassLoader m_system;

    private final ClassLoader m_common;

    private final LoggingManager m_logging;

    private final Logger m_logger;

    private final ModelFactory m_factory;

    private final Runtime m_runtime;

    private final long m_timeout;

    private final boolean m_secure;

    //--------------------------------------------------------------
    // mutable state
    //--------------------------------------------------------------

    private boolean m_trace;

    //--------------------------------------------------------------
    // constructor
    //--------------------------------------------------------------

   /**
    * Creation of a new system context.
    *
    * @param context the repository intial context
    * @param parent the parent context object (possibly null)
    * @param artifact an artifact reference to a plugin runtime factory
    * @param lifestyle an artifact reference to a plugin lifestyle factory
    * @param runtime the runtime class
    * @param logging the logging manager
    * @param base the base directory from which relative references 
    *   within a classpath or library directive shall be resolved
    * @param home the home directory
    * @param temp the temp directory
    * @param repository the application repository to be used when resolving 
    *   resource directives
    * @param category the kernel logging category name
    * @param trace flag indicating if internal logging is enabled
    * @param timeout a system wide default deployment timeout
    * @param security the security profiles
    */
    DefaultSystemContext( 
      InitialContext context, 
      Context parent, 
      Artifact artifact, 
      Artifact lifestyle, 
      Class runtime, 
      LoggingManager logging, 
      File base, 
      File home, 
      File temp, 
      File anchor, 
      Repository repository, 
      String category, 
      boolean trace, 
      long timeout, 
      boolean secure, 
      SecurityProfile[] security,
      Map grants ) throws SystemException
    {
        super( parent );

        assertNotNull( "context", context );
        assertNotNull( "base", base );
        assertNotNull( "anchor", anchor );
        assertNotNull( "repository", repository );
        assertNotNull( "logging", logging );
        assertNotNull( "category", category );
        assertNotNull( "security", security );
        assertNotNull( "grants", grants );

        if( !base.isDirectory() )
        {
            final String error = 
              REZ.getString( "system.error.base-not-a-directory", base  );
            throw new IllegalArgumentException( error );
        }

        m_base = base;
        m_home = home;
        m_temp = temp;
        m_anchor = anchor;
        m_trace = trace;
        m_repository = repository;
        m_logging = logging;
        m_timeout = timeout;

        m_logger = m_logging.getLoggerForCategory( category );
        m_system = SystemContext.class.getClassLoader();
        m_common = DeploymentModel.class.getClassLoader();

        m_secure = secure;

        if( m_secure )
        {
            if( System.getSecurityManager() == null )
            {
                System.setSecurityManager( new SecurityManager() );
            }
        }

        m_factory = new DefaultModelFactory( this, security, grants );

        //
        // use avalon-repository to load the runtime
        //

        if( null != runtime )
        {
            m_runtime = buildRuntimeInstance( context, runtime, lifestyle );
        }
        else if( null != artifact )
        {
            Class clazz = resolveRuntimeClass( context, artifact );
            m_runtime = buildRuntimeInstance( context, clazz, lifestyle );
        }
        else
        {
            m_runtime = null;
        }
    }

    //--------------------------------------------------------------
    // SystemContext
    //--------------------------------------------------------------

   /**
    * Return the enabled status of the code security policy.
    * @return the code security enabled status
    */
    public boolean isCodeSecurityEnabled()
    {
        return m_secure;
    }

   /**
    * Return the model factory.
    *
    * @return the factory
    */
    public ModelFactory getModelFactory()
    {
        return m_factory;
    }

   /**
    * Return the base directory from which relative classloader 
    * references may be resolved.
    *
    * @return the base directory
    */
    public File getBaseDirectory()
    {
        return m_base;
    }

   /**
    * Return the working directory from which containers may 
    * establish persistent content.
    *
    * @return the working directory
    */
    public File getHomeDirectory()
    {
        return m_home;
    }

   /**
    * Return the temporary directory from which a container 
    * may use to establish a transient content directory. 
    *
    * @return the temporary directory
    */
    public File getTempDirectory()
    {
        return m_temp;
    }

   /**
    * Return the anchor directory from which a container 
    * may use to resolve relative classpath references.
    *
    * @return the anchor directory
    */
    public File getAnchorDirectory()
    {
        return m_anchor;
    }

   /**
    * Return the system wide repository from which resource 
    * directives can be resolved.
    *
    * @return the repository
    */
    public Repository getRepository()
    {
        return m_repository;
    }

   /**
    * Return the API classloader.
    *
    * @return the system classloader
    */
    public ClassLoader getAPIClassLoader()
    {
        return m_common;
    }

   /**
    * Return the SPI classloader. 
    *
    * @return the system classloader
    */
    public ClassLoader getSPIClassLoader()
    {
        return m_system;
    }

   /**
    * Return the system trace flag.
    *
    * @return the trace flag
    */
    public boolean isTraceEnabled()
    {
        return m_trace;
    }

   /**
    * Set the system trace flag.
    *
    * @param trace the trace flag
    */
    public void setTraceEnabled( boolean trace )
    {
        m_trace = trace;
    }

   /**
    * Return the logging manager.
    *
    * @return the logging manager.
    */
    public LoggingManager getLoggingManager()
    {
        return m_logging;
    }

   /**
    * Return the system logging channel.
    *
    * @return the system logging channel
    */
    public Logger getLogger()
    {
        return m_logger;
    }

   /**
    * Return the default deployment phase timeout value.
    * @return the timeout value
    */
    public long getDefaultDeploymentTimeout()
    {
        return m_timeout;
    }

    //------------------------------------------------------------------
    // runtime operations
    //------------------------------------------------------------------

   /**
    * Request the commissioning of a runtime for a supplied deployment 
    * model.
    * @param model the deployment model 
    * @exception Exception of a commissioning error occurs
    */
    public void commission( DeploymentModel model ) throws Exception
    {
        getRuntime().commission( model );
    }

   /**
    * Request the decommissioning of a runtime for a supplied deployment 
    * model.
    * @param model the deployment model 
    * @exception Exception of a commissioning error occurs
    */
    public void decommission( DeploymentModel model )
    {
        getRuntime().decommission( model );
    }

   /**
    * Request resolution of an object from the runtime.
    * @param model the deployment model
    * @exception Exception if a deployment error occurs
    */
    public Object resolve( DeploymentModel model ) throws Exception
    {
        return getRuntime().resolve( model );
    }

   /**
    * Request resolution of an object from the runtime.
    * @param model the deployment model
    * @param proxy if TRUE the return value will be proxied if the 
    *   underlying component typoe suppports proxy representation 
    * @exception Exception if a deployment error occurs
    */
    public Object resolve( DeploymentModel model, boolean proxy ) throws Exception
    {
        return getRuntime().resolve( model, proxy );
    }

   /**
    * Request the release of an object from the runtime.
    * @param model the deployment model
    * @param instance the object to release
    * @exception Exception if a deployment error occurs
    */
    public void release( DeploymentModel model, Object instance )
    {
        getRuntime().release( model, instance );
    }

   /**
    * Prepare a string representation of an object for presentation.
    * @param object the object to parse
    * @return the presentation string
    */
    public String toString( Object object )
    {
        if( object == null ) return "";

        if( object instanceof String )
        {
            return processString( (String) object );
        }
        else
        {
            return processString( object.toString() );
        }
    }

   /**
    * Prepare a string representation of an object for presentation.
    * @param name the value to parse
    * @return the presentation string
    */
    public String processString( String name )
    {
        if( name == null ) return "";

        String str = name.replace( '\\', '/' );

        String base = getBaseDirectory().toString().replace( '\\', '/' );
        if( str.indexOf( base ) > -1 )
        {
            return getString( str, base, "${merlin.dir}" );
        }

        final String dir = 
          System.getProperty( "user.dir" ).replace( '\\', '/' );
        if( str.indexOf( dir ) > -1 )
        {
            return getString( str, dir, "${user.dir}" );
        }
        
        return name;
    }

   /**
    * Prepare a string representation of an object array for presentation.
    * @param objects the array of objects
    * @return the presentation string
    */
    public String toString( Object[] objects )
    {
        StringBuffer buffer = new StringBuffer();
        for( int i=0; i<objects.length; i++ )
        {
            if( i > 0 ) buffer.append( ";" );
            buffer.append( toString( objects[i] ) );
        }
        return buffer.toString();
    }

    //------------------------------------------------------------------
    // runtime operations
    //------------------------------------------------------------------

    private String getString( String name, String pattern, String replacement )
    {
        final int n = name.indexOf( pattern );
        if( n == -1 ) return name;
        if( name.startsWith( pattern ) )
        {
            return replacement + name.substring( pattern.length() );
        }
        else
        {
            String header = name.substring( 0, n );
            String tail = name.substring( n + pattern.length() );
            return header + replacement + tail; 
        }
    }

   /**
    * Return the runtime factory.
    *
    * @return the factory
    */
    private Runtime getRuntime()
    {
        if( null == m_runtime ) 
        {
            throw new IllegalStateException( "runtime" );
        }
        else
        {
            return m_runtime;
        }
    }

    private Class resolveRuntimeClass( InitialContext context, Artifact artifact )
      throws SystemException
    {
        if( null != artifact )
        {
            return getRuntimeClass( context, artifact );
        }
        else
        {
            return getRuntimeClass( 
              context, getDefaultRuntimeArtifact( context ) );
        }
    }

   /**
    * Get the runtime class referenced by the artifact.
    * @param context the repository initial context
    * @param artifact the factory artifact
    * @return the Runtime class
    */
    private Class getRuntimeClass( InitialContext context, Artifact artifact )
      throws SystemException
    {
        if( null == artifact )
        {
            throw new NullPointerException( "artifact" );
        }

        try
        {
            ClassLoader classloader = 
              DefaultSystemContext.class.getClassLoader();
            Builder builder = 
              context.newBuilder( classloader, artifact );
            return builder.getFactoryClass();
        }
        catch( Throwable e )
        {
            final String error = 
              REZ.getString( "system.error.load", artifact.toString() );
            throw new SystemException( error, e );
        }
    }

    private Artifact getDefaultRuntimeArtifact( InitialContext context )
    {
        Artifact[] artifacts = 
          context.getRepository().getCandidates( Runtime.class );
        if( artifacts.length > 0 ) return artifacts[0];

        final String error =
          "Initial context does not declare an artifact for the class: ["
          + Runtime.class.getName() + "].";
        throw new IllegalStateException( error );
    }

   /**
    * Build a runtime using a supplied class.
    *
    * @param context the initial context from which to bootstrap the runtime
    * @param clazz the runtime factory class
    * @param lifestyle the artifact referencing the lifestyle factory
    * @return a instance of the class
    * @exception SystemException if the class does not expose a public 
    *    constructor, or the constructor requires arguments that the 
    *    builder cannot resolve, or if a unexpected instantiation error 
    *    ooccurs
    */ 
    public Runtime buildRuntimeInstance( 
      InitialContext context, Class clazz, Artifact lifestyle ) 
      throws SystemException
    {
        if( null == clazz ) return null;

        Constructor[] constructors = clazz.getConstructors();
        if( constructors.length < 1 ) 
        {
            final String error = 
              REZ.getString( 
                "system.error.runtime.no-constructor", 
                clazz.getName() );
            throw new SystemException( error );
        }

        //
        // runtime class may has one constructor
        //

        Constructor constructor = constructors[0];
        Class[] classes = constructor.getParameterTypes();
        Object[] args = new Object[ classes.length ];
        for( int i=0; i<classes.length; i++ )
        {
            Class c = classes[i];
            if( SystemContext.class.isAssignableFrom( c ) )
            {
                args[i] = this;
            }
            else if( InitialContext.class.isAssignableFrom( c ) )
            {
                args[i] = context;
            }
            else if( Artifact.class.isAssignableFrom( c ) )
            {
                args[i] = lifestyle;
            }
            else
            {
                final String error = 
                  REZ.getString( 
                    "system.error.unrecognized-runtime-parameter", 
                    c.getName(),
                    clazz.getName() );
                throw new SystemException( error );
            }
        }

        //
        // instantiate the factory
        //

        return instantiateRuntime( constructor, args );
    }

   /**
    * Instantiation of a runtime using a supplied constructor 
    * and arguments.
    * 
    * @param constructor the runtime constructor
    * @param args the constructor arguments
    * @return the runtime instance
    * @exception SystemException if an instantiation error occurs
    */
    private Runtime instantiateRuntime( 
      Constructor constructor, Object[] args ) 
      throws SystemException
    {
        Class clazz = constructor.getDeclaringClass();
        try
        {
            return (Runtime) constructor.newInstance( args );
        }
        catch( Throwable e )
        {
            final String error = 
              REZ.getString( 
                "system.error.runtime-instantiation", 
                clazz.getName() );
            throw new SystemException( error, e );
        }
    }

    private void assertNotNull( String name, Object object )
    {
        if( null == object ) 
        {
            throw new NullPointerException( name );
        } 
    }
}
